tests/test_utils.py

```diff
@@ -221,58 +222,251 @@ def test_create_state_from_dict():
 
-    p1, p2 = (1, 3), (5, 3)
-    p3, p4 = (3, 7), (3, 2)
-    assert utils.intersects(p1, p2, p3, p4)
+    _, ax = plt.subplots(1, 1, figsize=(10, 10))
+    ax.set_xlim((-5, 5))
+    ax.set_ylim((-5, 5))
+    assert not utils.line_segment_intersects_circle(seg2, circ1, ax=ax)
+
+    # Uncomment for debugging.
+    # plt.savefig("/tmp/line_segment_circle_unit_test.png")
+
+
+def test_line_segment_rectangle_intersection():
+    """Tests for line_segment_intersects_rectangle()."""
+    seg1 = utils.LineSegment(-3, 0, 0, 0)
+    rect1 = utils.Rectangle(-1, -1, 2, 2, 0)
+    assert utils.geom2d_bodies_intersect(seg1, rect1)
+    assert utils.geom2d_bodies_intersect(rect1, seg1)
+
+    seg2 = utils.LineSegment(-3, 3, 4, 3)
+    assert not utils.geom2d_bodies_intersect(seg2, rect1)
+    assert not utils.geom2d_bodies_intersect(rect1, seg2)
+
+    seg3 = utils.LineSegment(0, -2, 1, -2.5)
+    assert not utils.geom2d_bodies_intersect(seg3, rect1)
+    assert not utils.geom2d_bodies_intersect(rect1, seg3)
+
+    seg4 = utils.LineSegment(0, -3, 0, -4)
+    assert not utils.geom2d_bodies_intersect(seg4, rect1)
+    assert not utils.geom2d_bodies_intersect(rect1, seg4)
 
-    p1, p2 = (2, 5), (7, 6)
-    p3, p4 = (2, 6), (7, 7)
-    assert not utils.intersects(p1, p2, p3, p4)
 
-    p1, p2 = (1, 1), (3, 3)
-    p3, p4 = (2, 2), (4, 4)
-    assert not utils.intersects(p1, p2, p3, p4)
+def test_rectangle_circle_intersection():
+    """Tests for rectangle_intersects_circle()."""
+    rect1 = utils.Rectangle(x=0, y=0, width=4, height=3, theta=0)
+    circ1 = utils.Circle(x=0, y=0, radius=1)
+    assert utils.geom2d_bodies_intersect(rect1, circ1)
+    assert utils.geom2d_bodies_intersect(circ1, rect1)
+
+    circ2 = utils.Circle(x=1, y=1, radius=0.5)
+    assert utils.geom2d_bodies_intersect(rect1, circ2)
+    assert utils.geom2d_bodies_intersect(circ2, rect1)
 
-    p1, p2 = (1, 1), (3, 3)
-    p3, p4 = (1, 1), (6.7, 7.4)
-    assert not utils.intersects(p1, p2, p3, p4)
+    rect2 = utils.Rectangle(x=1, y=1, width=1, height=1, theta=0)
+    assert not utils.geom2d_bodies_intersect(rect2, circ1)
+    assert not utils.geom2d_bodies_intersect(circ1, rect2)
 
+    circ3 = utils.Circle(x=0, y=0, radius=100)
+    assert utils.geom2d_bodies_intersect(rect1, circ3)
+    assert utils.geom2d_bodies_intersect(circ3, rect1)
+    assert utils.geom2d_bodies_intersect(rect2, circ3)
+    assert utils.geom2d_bodies_intersect(circ3, rect2)
 
-def test_overlap():
-    """Tests for overlap()."""
-    l1, r1 = (1, 7), (3, 1)
-    l2, r2 = (2, 10), (7, 3)
-    assert utils.overlap(l1, r1, l2, r2)
 
-    l1, r1 = (1, 7), (3, 1)
-    l2, r2 = (1, 8), (6, 1)
-    assert utils.overlap(l1, r1, l2, r2)
+def test_geom2d_bodies_intersect():
+    """Tests for geom2d_bodies_intersect()."""
 
-    l1, r1 = (1, 7), (5, 1)
-    l2, r2 = (2, 4), (4, 2)
-    assert utils.overlap(l1, r1, l2, r2)
+    class _MockGeom2DBody(utils._Geom2DBody):  # pylint: disable=protected-access
```

Comment:
you can get rid of this pylint disable by subclassing from one of the public ones instead (Circle, Rectangle)

Reply:
I can't do that because then the `isinstance` checks will get caught

Reply:
okay I think I found a workaround (pushed)